service/history/replication/executable_task.go

```diff
@@ -435,6 +448,106 @@ FilterLoop:
 	return namespaceEntry.Name().String(), shouldProcessTask, nil
 }
 
+func (e *ExecutableTaskImpl) Import(
+	ctx context.Context,
+	remoteCluster string,
+	workflowKey definition.WorkflowKey,
+	startEventId int64,
+	startEventVersion int64,
+	endEventId int64,
+	endEventVersion int64,
+) error {
+	e.MetricsHandler.Counter(metrics.ClientRequests.GetMetricName()).Record(
+		1,
+		metrics.OperationTag(e.metricsTag+"Import"),
+	)
+	startTime := time.Now().UTC()
+	defer func() {
+		e.MetricsHandler.Timer(metrics.ClientLatency.GetMetricName()).Record(
+			time.Since(startTime),
+			metrics.OperationTag(e.metricsTag+"Import"),
+		)
+	}()
+
+	historyIterator := e.ProcessToolBox.HistoryPaginatedFetcher.GetSingleWorkflowHistoryPaginatedIterator(
+		ctx,
+		remoteCluster,
+		namespace.ID(workflowKey.NamespaceID),
+		workflowKey.WorkflowID,
+		workflowKey.RunID,
+		startEventId,
+		startEventVersion,
+		endEventId,
+		endEventVersion,
+	)
+
+	shardContext, err := e.ShardController.GetShardByNamespaceWorkflow(
+		namespace.ID(workflowKey.NamespaceID),
+		workflowKey.WorkflowID,
+	)
+	if err != nil {
+		return err
+	}
+	engine, err := shardContext.GetEngine(ctx)
+	if err != nil {
+		return err
+	}
+
+	var token []byte
+	blobs := []*commonpb.DataBlob{}
+	blobSize := 0
+	var versionHistory *historyspb.VersionHistory
+
+	for historyIterator.HasNext() {
+		batch, err := historyIterator.Next()
+		if err != nil {
+			e.Logger.Error("failed to get history events",
+				tag.WorkflowNamespaceID(workflowKey.NamespaceID),
+				tag.WorkflowID(workflowKey.WorkflowID),
+				tag.WorkflowRunID(workflowKey.RunID),
+				tag.Error(err))
+			return err
+		}
+		if versionHistory != nil && versionHistory != batch.VersionHistory {
```

Comment:
what if during this pagination, workflow on src cell make progress?

Reply:
Nice catch and thank you for your suggest of checking the branch instead of comparing equality.